File: Python/DailyLeetCode/shortestDistance.py
from collections import defaultdict

class Solution:
    def shortestDistanceAfterQueries(self, n, queries):
        graph = defaultdict(list)
        distances = list(range(n))
        result = []

        for i in range(1, n):
            graph[i].append(i - 1)

        def update_distances(node):
            new_dist = distances[node] + 1
            for neighbor in graph[node]:
                if distances[neighbor] <= new_dist:
                    continue
                distances[neighbor] = new_dist
                update_distances(neighbor)

        for source, target in queries:
            graph[target].append(source)
            distances[source] = min(distances[source], distances[target] + 1)
            update_distances(source)
            result.append(distances[0])

        return result

File: Java/DailyLeetCode/shortestDistance.java
import java.util.*;
class shortDistance {
    public List<Integer> shortestDistanceAfterQueries(int n, List<int[]> queries) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        int[] distances = new int[n];
        Arrays.fill(distances, n - 1);

        for (int i = 1; i < n; i++) {
            graph.get(i).add(i - 1);
        }

        List<Integer> answer = new ArrayList<>();
        for (int[] query : queries) {
            int source = query[0], target = query[1];
            graph.get(target).add(source);
            distances[source] = Math.min(distances[source], distances[target] + 1);
            updateDistances(graph, source, distances);
            answer.add(distances[0]);
        }

        return answer;
    }

    private void updateDistances(List<List<Integer>> graph, int current, int[] distances) {
        int newDist = distances[current] + 1;
        for (int neighbor : graph.get(current)) {
            if (distances[neighbor] <= newDist) continue;
            distances[neighbor] = newDist;
            updateDistances(graph, neighbor, distances);
        }
    }
}